### PR TITLE
manager: Add a test case in test_node_config_annotation_invalid() for duplicate disk name

### PR DESCRIPTION
**Pull request by boknowswiki**
This test case is to cover no duplicate disk names in node default disk
config annotation

https://github.com/longhorn/longhorn/issues/1219

Add another test case for multiple paths in the same filesystem

https://github.com/longhorn/longhorn/issues/1293

Signed-off-by: Bo Tao <bo.tao@rancher.com>

### CODE REVIEW

**Comment by boknowswiki** [comment]
Verified the longhorn-manager log has the correct error message
`[longhorn-manager-b559j] time="2020-05-02T06:43:55Z" level=warning msg="Kubernetes node: invalid annotation node.longhorn.io/default-disks-config: config: the disk name same-name has duplicated" `

File: manager/integration/tests/test_node.py [hunk 1]
```diff
@@ -1647,6 +1653,54 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
...
+        "metadata": {
+            "labels": {
+                CREATE_DEFAULT_DISK_LABEL: None
+            },
+            "annotations": {
+                DEFAULT_DISK_CONFIG_ANNOTATION: None,
+            }
+        }
+    })
+
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1703]: The created host disk needs to be cleaned up as well. Not sure it's covered by the fixture already.

Check test_nodes.py for examples.

File: manager/integration/tests/test_node.py [hunk 2]
```diff
@@ -1507,14 +1507,76 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
...
     # intervene the node controller.
+
+    # Case1.1: Multiple paths on the same filesystem is invalid disk
+    # annotation, Longhorn shouldn't apply this.
+
+    # make a clean condition for test to start.
+    cleanup_node_disks(client, node_name)
+    client.update(node, tags=[])
+
+    core_api.patch_node(node_name, {
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1525]: `reset_disk_and_tag_annotations` fixture already did this.

File: manager/integration/tests/test_node.py [hunk 3]
```diff
@@ -1507,14 +1507,76 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
...
+                    '"storageReserved":1024,"name":"root-name"},' +
+                    '{"path":"' + DEFAULT_DISK_PATH +
+                    '","allowScheduling":false,' +
+                    '"storageReserved": 1024,' +
+                    '"name":"default-name"}]'
+            }
+        }
+    })
+
+    time.sleep(NODE_UPDATE_WAIT_INTERVAL)
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1557]: Shouldn't need to wait here since this is for node annotation to be reflected in Longhorn, not for Kubernetes node to be updated. Once the API returns, the Kubernetes node should be updated already.

File: manager/integration/tests/test_node.py [hunk 4]
```diff
@@ -1507,14 +1507,60 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
        "Create Default Disk on Labeled Nodes" in settings.
     3. The node tag or disks won't be updated.
     """
-    setting = client.by_id_setting(SETTING_CREATE_DEFAULT_DISK_LABELED_NODES)
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1504]: Why moving this?

File: manager/integration/tests/test_node.py [hunk 5]
```diff
@@ -1507,14 +1507,60 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
...
-    client.update(setting, value="true")
 
     nodes = client.list_node().data
     node_name = nodes[0].id
+    node = client.by_id_node(node_name)
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1513]: When is this variable needed?

File: manager/integration/tests/test_node.py [hunk 6]
```diff
@@ -1507,14 +1507,60 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
...
+        "metadata": {
+            "labels": {
+                CREATE_DEFAULT_DISK_LABEL:
+                    CREATE_DEFAULT_DISK_LABEL_VALUE_CONFIG
+            },
+            "annotations": {
+                DEFAULT_DISK_CONFIG_ANNOTATION:
+                    '[{"path":"/root","allowScheduling":false,' +
+                    '"storageReserved":1024,"name":"root-name"},' +
+                    '{"path":"' + DEFAULT_DISK_PATH +
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1535]: Don't need to use `DEFAULT_DISK_PATH`. We can just do `/root` and `/home`. Both are likely on the root fs.

File: manager/integration/tests/test_node.py [hunk 7]
```diff
@@ -1507,14 +1507,60 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
...
+    setting = client.by_id_setting(SETTING_CREATE_DEFAULT_DISK_LABELED_NODES)
+    client.update(setting, value="true")
+    # Longhorn shouldn't apply the invalid disk annotation.
+    time.sleep(NODE_UPDATE_WAIT_INTERVAL)
+    node = client.by_id_node(node_name)
+    assert len(node.disks) == 0
+    assert not node.tags
+
+    # Cleanup invalid annotation
+    core_api.patch_node(node_name, {
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1552]: We probably don't need this cleanup, as long as no disk was created. The next test case will override it anyway.

File: manager/integration/tests/test_node.py [hunk 8]
```diff
@@ -1647,6 +1653,56 @@ def test_node_config_annotation_invalid(client, core_api,  # NOQA
...
+            }
+        }
+    })
+
+    # same disk name shouldn't be applied to Longhorn.
+    time.sleep(NODE_UPDATE_WAIT_INTERVAL)
+    node = client.by_id_node(lht_hostId)
+    assert len(node.disks) == 0
+
+    # do cleanup.
```
Comment (yasker) [inline manager/integration/tests/test_node.py:1738]: Don't need to do this. `reset_disk_and_tag_annotations` should has covered this already.

**Comment by boknowswiki** [comment]
Verified commit 1c6b734 with local built longhorn-manager:

[longhorn-manager-t87z5] time="2020-05-05T18:25:33Z" level=warning msg="Kubernetes node: invalid annotation node.longhorn.io/default-disks-config: config: the disk /home is the samefile system with /root, fsid 58fe937c58377e45"
